### ISSUE TITLE
What's the proper way to align text and image embeddings?

### ISSUE THREAD

**Issue by skylinezum** [apple/ml-fastvlm#20]
Hi, I am having trouble aligning the text semantic embeddings with image embeddings that I extract from the model. I use the mm_projector to project the image embeddings and then normalize, but the text input and the images matched are quite off. The image -> image comparison does very well as the semantic meaning is retained after projecting, but it seems that there is some misalignment with text.

Does the projection size look correct or is it possible that my text embeddings are taken from the wrong layer?
It would also be really cool if a method could be added to the model that could directly extract an aligned embedding from either text or image.

**Image Embeddings Shape after Projection and Normalization**
Image tensor shape: torch.Size([1, 3, 1024, 1024]), dtype: torch.float16
2025-05-16 00:08:02,185 - INFO - Raw image embeddings shape: torch.Size([1, 256, 3072])
2025-05-16 00:08:02,194 - INFO - Projected embeddings shape: torch.Size([1, 256, 896])
2025-05-16 00:08:02,373 - INFO - Embeddings normalized
2025-05-16 00:08:02,390 - INFO - Successfully extracted embeddings with shape torch.Size([1, 896])

**Text Embeddings Shape after Normalization**
2025-05-16 00:08:36,790 - INFO - Text embedding shape: torch.Size([1, 896])

**Comment by pavank-apple** [comment]
@skylinezum 
I am not sure how you are extracting the text embedding, since this model was not specifically trained for that. For 0.5B model, the tensor shapes look correct to me. It makes sense that image -> image comparison works since the backbone was CLIP pre-trained. But the text encoder used during CLIP pre-training was discarded. Are you looking for the CLIP pre-trained backbone? 
If not, then I am afraid you would have to do something like [ColPali](https://arxiv.org/abs/2407.01449) to be able to extract reliable embeddings from this VLM.

**Comment by skylinezum** [comment]
Thanks for the quick reply, Actually yes, I think the CLIP pre-trained backbone would be more useful for what I want. Is that available somewhere? I'm trying to text -> image and image -> image search using different VLM and wanted to give this a shot.

**Comment by pavank-apple** [comment]
@skylinezum 
Here is our prior work [MobileCLIP](https://github.com/apple/ml-mobileclip), hopefully that can be of help. That repo also contains CoreML models and a demo app to run CLIP models efficiently on device, text -> image & image -> image search is better performed by embedding models like CLIP.

**Comment by skylinezum** [comment]
@pavank-apple Thank you for pointing me in a better direction.
